File: Kalorie/Kalorie/Features/FoodQuantity/FoodQuantityViewModel.swift
//
//  FoodQuantityViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 24.07.2026.
//

import Foundation

enum FoodQuantityUnit: Hashable {
    case hundredGrams
    case grams
    case portion(FoodPortionDomain)

    var gramsPerUnit: Double {
        switch self {
        case .hundredGrams: return 100
        case .grams: return 1
        case .portion(let portion): return portion.grams
        }
    }
}

final class FoodQuantityViewModel: ObservableObject, FavouriteToggling, FoodItemReporting {

    // MARK: - Properties

    @Published var unit: FoodQuantityUnit
    @Published var quantity: Double
    @Published private(set) var state: LoadingState<Void> = .idle
    @Published var alertItem: AlertItem?
    @Published var isFavourite: Bool
    @Published var isTogglingFavourite = false
    @Published private(set) var personalPortions: [FoodPortionDomain] = []
    @Published var isPersonalPortionsManagerPushed = false
    @Published var portionDrafts = [FoodPortionDraft(name: "", gramsText: "")]
    @Published private(set) var showPortionCheckmark = false
    @Published private(set) var mealTypes: [MealTypeDomain]
    @Published var selectedMealTypeId: String?
    @Published var hasReportedCurrentItem = false
    @Published var isSubmittingReport = false
    @Published var isReportReasonAlertVisible = false
    @Published var reportReasonText = ""

    let item: FoodItemDomain
    private var meal: MyCreatedMealDomain?
    private let saveFoodConsumed: any SaveFoodConsumedUseCaseProtocol
    private let fetchMealTypes: any FetchMealTypesUseCaseProtocol
    private let addFavouriteFood: any AddFavouriteFoodUseCaseProtocol
    private let removeFavouriteFood: any RemoveFavouriteFoodUseCaseProtocol
    private let fetchFoodItemPersonalPortions: any FetchFoodItemPersonalPortionsUseCaseProtocol
    private let saveFoodItemPersonalPortions: any SaveFoodItemPersonalPortionsUseCaseProtocol
    private let fetchMyFoodItemReport: any FetchMyFoodItemReportUseCaseProtocol
    private let submitFoodItemReport: any SubmitFoodItemReportUseCaseProtocol
    private let updateMyCreatedMeal: any UpdateMyCreatedMealUseCaseProtocol
    private let selectedDate: Date
    private let onSaved: () -> Void
    private let onFavouriteChanged: (String, Bool) -> Void
    private let onMealUpdated: (MyCreatedMealDomain) -> Void
    private var hasUserSelectedUnit = false
    private var hasUserSelectedMealType = false

    var grams: Double { quantity * unit.gramsPerUnit }

    private var scaledMacros: ScaledMacros { item.scaled(toGrams: grams) }

    var scaledCalories: Int { scaledMacros.calories }
    var scaledProtein: Double { scaledMacros.protein }
    var scaledCarbohydrate: Double { scaledMacros.carbohydrate }
    var scaledCarbohydrateSugar: Double { scaledMacros.carbohydrateSugar }
    var scaledFat: Double { scaledMacros.fat }
    var scaledFatSaturated: Double? { scaledMacros.fatSaturated }
    var scaledFiber: Double? { scaledMacros.fiber }
    var scaledSalt: Double { scaledMacros.salt }

    var arePortionDraftsComplete: Bool {
        portionDrafts.allSatisfy { draft in
            !draft.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !draft.gramsText.isEmpty
        }
    }

    var canSavePortionDrafts: Bool {
        portionDrafts.contains { draft in
            !draft.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !draft.gramsText.isEmpty
        }
    }

    var isPersonalPortionsAvailable: Bool { item.kind == .catalogue || meal != nil }
    var canReportIncorrectData: Bool { item.kind == .catalogue }

    var unitOptions: [FoodQuantityUnit] {
        let ownPortions = meal == nil ? item.portions : []
        return personalPortions.map(FoodQuantityUnit.portion) + ownPortions.map(FoodQuantityUnit.portion) + [.grams, .hundredGrams]
    }

    // MARK: - Init

    init(
        item: FoodItemDomain,
        saveFoodConsumed: any SaveFoodConsumedUseCaseProtocol,
        fetchMealTypes: any FetchMealTypesUseCaseProtocol,
        selectedDate: Date,
        mealTypes: [MealTypeDomain],
        isFavourite: Bool,
        addFavouriteFood: any AddFavouriteFoodUseCaseProtocol,
        removeFavouriteFood: any RemoveFavouriteFoodUseCaseProtocol,
        fetchFoodItemPersonalPortions: any FetchFoodItemPersonalPortionsUseCaseProtocol,
        saveFoodItemPersonalPortions: any SaveFoodItemPersonalPortionsUseCaseProtocol,
        fetchMyFoodItemReport: any FetchMyFoodItemReportUseCaseProtocol,
        submitFoodItemReport: any SubmitFoodItemReportUseCaseProtocol,
        meal: MyCreatedMealDomain?,
        updateMyCreatedMeal: any UpdateMyCreatedMealUseCaseProtocol,
        onSaved: @escaping () -> Void,
        onMealUpdated: @escaping (MyCreatedMealDomain) -> Void,
        onFavouriteChanged: @escaping (String, Bool) -> Void,
        quantity: Double = 1,
        unit: FoodQuantityUnit = .hundredGrams
    ) {
        self.item = item
        self.saveFoodConsumed = saveFoodConsumed
        self.fetchMealTypes = fetchMealTypes
        self.selectedDate = selectedDate
        self.mealTypes = mealTypes
        self.selectedMealTypeId = mealTypes.mealType(at: selectedDate)?.id
        self.isFavourite = isFavourite
        self.addFavouriteFood = addFavouriteFood
        self.removeFavouriteFood = removeFavouriteFood
        self.fetchFoodItemPersonalPortions = fetchFoodItemPersonalPortions
        self.saveFoodItemPersonalPortions = saveFoodItemPersonalPortions
        self.fetchMyFoodItemReport = fetchMyFoodItemReport
        self.submitFoodItemReport = submitFoodItemReport
        self.meal = meal
        self.updateMyCreatedMeal = updateMyCreatedMeal
        self.personalPortions = meal?.portions ?? []
        self.onSaved = onSaved
        self.onFavouriteChanged = onFavouriteChanged
        self.onMealUpdated = onMealUpdated
        self.quantity = quantity
        self.unit = unit
    }

    // MARK: - Functions

    static func defaultUnit(for item: FoodItemDomain) -> FoodQuantityUnit {
        item.portions.first.map(FoodQuantityUnit.portion) ?? .grams
    }

    @MainActor
    func onAppear() async {
        if canReportIncorrectData {
            await loadReportState(barcode: item.id, fetchMyFoodItemReport: fetchMyFoodItemReport)
        }
        guard item.kind == .catalogue else { return }
        do {
            personalPortions = try await fetchFoodItemPersonalPortions(barcode: item.id)
            if
                !hasUserSelectedUnit,
                let firstPersonalPortion = personalPortions.first
            {
                quantity = 1
                unit = .portion(firstPersonalPortion)
            }
        } catch {
            Log.warning(error, category: Constants.LogCategory.foodQuantity)
        }
    }

    @MainActor
    func onReportSubmitted() async {
        await onReportSubmitted(barcode: item.id, submitFoodItemReport: submitFoodItemReport)
    }

    func onUnitSelected(_ newUnit: FoodQuantityUnit) {
        hasUserSelectedUnit = true
        let currentGrams = quantity * unit.gramsPerUnit
        unit = newUnit
        quantity = currentGrams / newUnit.gramsPerUnit
    }

    func onMealTypeSelected(_ mealTypeId: String) {
        hasUserSelectedMealType = true
        selectedMealTypeId = mealTypeId
    }

    func onPortionsManagerOpened() {
        if case .portion = unit {
            portionDrafts = [FoodPortionDraft(name: "", gramsText: "")]
        } else {
            portionDrafts = [FoodPortionDraft(name: "", gramsText: grams.formattedTrimmed())]
        }
    }

    func onAddPortionDraftTapped() {
        portionDrafts.append(FoodPortionDraft(name: "", gramsText: ""))
    }

    func onDeletePortionDraft(_ draft: FoodPortionDraft) {
        portionDrafts.removeAll { $0.id == draft.id }
        if portionDrafts.isEmpty {
            portionDrafts = [FoodPortionDraft(name: "", gramsText: "")]
        }
    }

    @MainActor
    func onSavePersonalPortions() async {
        let filledDrafts = portionDrafts.filter { draft in
            !draft.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || !draft.gramsText.isEmpty
        }
        guard !filledDrafts.isEmpty else { return }
        var newPortions: [FoodPortionDomain] = []
        for draft in filledDrafts {
            let grams = Double(draft.gramsText.replacingOccurrences(of: ",", with: ".")) ?? 0
            if let error = FoodPortionValidation.validate(name: draft.name, grams: grams) {
                alertItem = AlertItem(title: error.alertTitle)
                return
            }
            newPortions.append(FoodPortionDomain(name: draft.name, grams: grams))
        }
        let original = personalPortions
        personalPortions = original + newPortions
        do {
            try await persistPersonalPortions()
            portionDrafts = [FoodPortionDraft(name: "", gramsText: "")]
            showPortionCheckmark = true
            try? await Task.sleep(for: .seconds(2))
            showPortionCheckmark = false
        } catch let error as FoodPortionError {
            Log.error(error, category: Constants.LogCategory.foodQuantity)
            personalPortions = original
            alertItem = AlertItem(title: error.alertTitle)
        } catch {
            Log.error(error, category: Constants.LogCategory.foodQuantity)
            personalPortions = original
            alertItem = AlertItem(title: L10n.MyPortions.errorSaveFailed)
        }
    }

    @MainActor
    func onDeletePersonalPortion(_ portion: FoodPortionDomain) async {
        let original = personalPortions
        let originalUnit = unit
        let originalQuantity = quantity
        personalPortions.removeAll { $0 == portion }
        if !unitOptions.contains(unit) {
            quantity = grams
            unit = .grams
        }
        do {
            try await persistPersonalPortions()
        } catch {
            Log.error(error, category: Constants.LogCategory.foodQuantity)
            personalPortions = original
            unit = originalUnit
            quantity = originalQuantity
            alertItem = AlertItem(title: L10n.MyPortions.errorDeleteFailed)
        }
    }

    @MainActor
    private func persistPersonalPortions() async throws {
        guard let meal else {
            try await saveFoodItemPersonalPortions(barcode: item.id, portions: personalPortions)
            return
        }
        let updated = MyCreatedMealDomain(
            id: meal.id,
            name: meal.name,
            ingredients: meal.ingredients,
            createdAt: meal.createdAt,
            updatedAt: meal.updatedAt,
            portions: personalPortions
        )
        try await updateMyCreatedMeal(updated)
        self.meal = updated
        onMealUpdated(updated)
    }

    @MainActor
    func onFavouriteToggled() async {
        let itemId = item.id
        await toggleFavourite(
            item: item,
            removalId: itemId,
            addFavouriteFood: addFavouriteFood,
            removeFavouriteFood: removeFavouriteFood
        ) { [weak self] newValue in
            self?.onFavouriteChanged(itemId, newValue)
        }
    }

    @MainActor
    func onConfirm() async {
        guard !state.isLoading else { return }
        guard grams > 0 else {
            alertItem = AlertItem(title: L10n.FoodQuantity.errorInvalidQuantity)
            return
        }
        state = .loading
        defer { state = .loaded }
        do {
            do {
                mealTypes = try await fetchMealTypes()
            } catch {
                Log.warning(error, category: Constants.LogCategory.foodQuantity)
            }
            let mealTypeId: String?
            if hasUserSelectedMealType {
                guard
                    let selectedMealTypeId,
                    mealTypes.contains(where: { $0.id == selectedMealTypeId })
                else {
                    alertItem = AlertItem(title: L10n.Common.errorUnknown)
                    return
                }
                mealTypeId = selectedMealTypeId
            } else {
                mealTypeId = mealTypes.mealType(at: selectedDate)?.id
            }
            try await saveFoodConsumed(item, grams: grams, date: selectedDate, mealTypeId: mealTypeId)
            onSaved()
        } catch {
            Log.error(error, category: Constants.LogCategory.foodQuantity)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/foodquantity/FoodQuantityViewModel.kt
package antoni.kalorie.features.foodquantity

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.components.FoodPortionDraft
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemKind
import antoni.kalorie.core.models.FoodPortionDomain
import antoni.kalorie.core.models.FoodPortionError
import antoni.kalorie.core.models.FoodPortionValidation
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.models.MyCreatedMealDomain
import antoni.kalorie.core.models.ScaledMacros
import antoni.kalorie.core.models.mealType
import antoni.kalorie.core.models.scaled
import antoni.kalorie.core.usecases.AddFavouriteFoodUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodItemPersonalPortionsUseCaseProtocol
import antoni.kalorie.core.usecases.FetchMealTypesUseCaseProtocol
import antoni.kalorie.core.usecases.RemoveFavouriteFoodUseCaseProtocol
import antoni.kalorie.core.usecases.SaveFoodConsumedUseCaseProtocol
import antoni.kalorie.core.usecases.SaveFoodItemPersonalPortionsUseCaseProtocol
import antoni.kalorie.core.usecases.UpdateMyCreatedMealUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.FavouriteToggling
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import antoni.kalorie.core.utils.isLoading
import java.time.Instant
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

sealed interface FoodQuantityUnit {
    data object HundredGrams : FoodQuantityUnit
    data object Grams : FoodQuantityUnit
    data class Portion(val portion: FoodPortionDomain) : FoodQuantityUnit

    // MARK: - Properties

    val gramsPerUnit: Double
        get() = when (this) {
            HundredGrams -> 100.0
            Grams -> 1.0
            is Portion -> portion.grams
        }
}

class FoodQuantityViewModel(
    val item: FoodItemDomain,
    private val saveFoodConsumed: SaveFoodConsumedUseCaseProtocol,
    private val fetchMealTypes: FetchMealTypesUseCaseProtocol,
    private val selectedDate: Instant,
    mealTypes: List<MealTypeDomain>,
    isFavourite: Boolean,
    private val addFavouriteFood: AddFavouriteFoodUseCaseProtocol,
    private val removeFavouriteFood: RemoveFavouriteFoodUseCaseProtocol,
    private val fetchFoodItemPersonalPortions: FetchFoodItemPersonalPortionsUseCaseProtocol,
    private val saveFoodItemPersonalPortions: SaveFoodItemPersonalPortionsUseCaseProtocol,
    private var meal: MyCreatedMealDomain?,
    private val updateMyCreatedMeal: UpdateMyCreatedMealUseCaseProtocol,
    private val onSaved: () -> Unit,
    private val onMealUpdated: (MyCreatedMealDomain) -> Unit,
    private val onFavouriteChanged: (String, Boolean) -> Unit,
    quantity: Double = 1.0,
    unit: FoodQuantityUnit = FoodQuantityUnit.HundredGrams,
) : ViewModel(), FavouriteToggling {

    // MARK: - Properties

    val unit = MutableStateFlow(unit)
    val quantity = MutableStateFlow(quantity)
    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Unit>> = _state
    override val alertItem = MutableStateFlow<AlertItem?>(null)
    override val isFavourite = MutableStateFlow(isFavourite)
    override val isTogglingFavourite = MutableStateFlow(false)
    private val _mealTypes = MutableStateFlow(mealTypes)
    val mealTypes: StateFlow<List<MealTypeDomain>> = _mealTypes
    val selectedMealTypeId = MutableStateFlow(mealTypes.mealType(selectedDate)?.id)
    private val _personalPortions = MutableStateFlow(meal?.portions ?: emptyList())
    val personalPortions: StateFlow<List<FoodPortionDomain>> = _personalPortions
    val isPersonalPortionsManagerPushed = MutableStateFlow(false)
    val portionDrafts = MutableStateFlow(listOf(FoodPortionDraft.blank))
    private val _showPortionCheckmark = MutableStateFlow(false)
    val showPortionCheckmark: StateFlow<Boolean> = _showPortionCheckmark
    private var hasUserSelectedUnit = false
    private var hasUserSelectedMealType = false

    val grams: Double
        get() = quantity.value * unit.value.gramsPerUnit

    private val scaledMacros: ScaledMacros
        get() = item.scaled(toGrams = grams)

    val scaledCalories: Int get() = scaledMacros.calories
    val scaledProtein: Double get() = scaledMacros.protein
    val scaledCarbohydrate: Double get() = scaledMacros.carbohydrate
    val scaledCarbohydrateSugar: Double get() = scaledMacros.carbohydrateSugar
    val scaledFat: Double get() = scaledMacros.fat
    val scaledFatSaturated: Double? get() = scaledMacros.fatSaturated
    val scaledFiber: Double? get() = scaledMacros.fiber
    val scaledSalt: Double get() = scaledMacros.salt

    val arePortionDraftsComplete: Boolean
        get() = portionDrafts.value.all { it.isComplete }

    val canSavePortionDrafts: Boolean
        get() = portionDrafts.value.any { it.isComplete }

    val isPersonalPortionsAvailable: Boolean
        get() = item.kind == FoodItemKind.CATALOGUE || meal != null

    val unitOptions: List<FoodQuantityUnit>
        get() = (_personalPortions.value + (if (meal == null) item.portions else emptyList())).map { FoodQuantityUnit.Portion(it) } +
            listOf(FoodQuantityUnit.Grams, FoodQuantityUnit.HundredGrams)

    // MARK: - Functions

    suspend fun onAppear() {
        if (item.kind != FoodItemKind.CATALOGUE) return
        try {
            _personalPortions.value = fetchFoodItemPersonalPortions(item.id)
            val firstPersonalPortion = _personalPortions.value.firstOrNull()
            if (!hasUserSelectedUnit && firstPersonalPortion != null) {
                quantity.value = 1.0
                unit.value = FoodQuantityUnit.Portion(firstPersonalPortion)
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.FOOD_QUANTITY)
        }
    }

    fun onUnitSelected(newUnit: FoodQuantityUnit) {
        hasUserSelectedUnit = true
        val currentGrams = quantity.value * unit.value.gramsPerUnit
        unit.value = newUnit
        quantity.value = currentGrams / newUnit.gramsPerUnit
    }

    fun onMealTypeSelected(mealTypeId: String) {
        hasUserSelectedMealType = true
        selectedMealTypeId.value = mealTypeId
    }

    fun onPortionsManagerOpened() {
        portionDrafts.value = listOf(FoodPortionDraft.blank)
    }

    fun onAddPortionDraftTapped() {
        portionDrafts.value = portionDrafts.value + FoodPortionDraft.blank
    }

    fun onDeletePortionDraft(draft: FoodPortionDraft) {
        portionDrafts.value = portionDrafts.value.filter { it.id != draft.id }.ifEmpty { listOf(FoodPortionDraft.blank) }
    }

    suspend fun onSavePersonalPortions() {
        val filledDrafts = portionDrafts.value.filter { it.name.isNotBlank() || it.gramsText.isNotEmpty() }
        if (filledDrafts.isEmpty()) return
        val newPortions = mutableListOf<FoodPortionDomain>()
        for (draft in filledDrafts) {
            val grams = draft.gramsText.replace(',', '.').toDoubleOrNull() ?: 0.0
            val error = FoodPortionValidation.validate(name = draft.name, grams = grams)
            if (error != null) {
                alertItem.value = AlertItem(titleRes = error.alertTitleRes)
                return
            }
            newPortions.add(FoodPortionDomain(name = draft.name, grams = grams))
        }
        val original = _personalPortions.value
        _personalPortions.value = original + newPortions
        try {
            persistPersonalPortions()
            portionDrafts.value = listOf(FoodPortionDraft.blank)
            _showPortionCheckmark.value = true
            delay(CHECKMARK_DURATION_MILLIS)
            _showPortionCheckmark.value = false
        } catch (error: CancellationException) {
            throw error
        } catch (error: FoodPortionError) {
            Log.error(error, Constants.LogCategory.FOOD_QUANTITY)
            _personalPortions.value = original
            alertItem.value = AlertItem(titleRes = error.alertTitleRes)
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.FOOD_QUANTITY)
            _personalPortions.value = original
            alertItem.value = AlertItem(titleRes = R.string.myPortions_error_saveFailed)
        }
    }

    suspend fun onDeletePersonalPortion(portion: FoodPortionDomain) {
        val original = _personalPortions.value
        val originalUnit = unit.value
        val originalQuantity = quantity.value
        _personalPortions.value = original.filter { it != portion }
        if (unitOptions.none { it == unit.value }) {
            quantity.value = grams
            unit.value = FoodQuantityUnit.Grams
        }
        try {
            persistPersonalPortions()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.FOOD_QUANTITY)
            _personalPortions.value = original
            unit.value = originalUnit
            quantity.value = originalQuantity
            alertItem.value = AlertItem(titleRes = R.string.myPortions_error_deleteFailed)
        }
    }

    private suspend fun persistPersonalPortions() {
        val currentMeal = meal
        if (currentMeal == null) {
            saveFoodItemPersonalPortions(item.id, _personalPortions.value)
            return
        }
        val updated = currentMeal.copy(portions = _personalPortions.value)
        updateMyCreatedMeal(updated)
        meal = updated
        onMealUpdated(updated)
    }

    suspend fun onFavouriteToggled() {
        val itemId = item.id
        toggleFavourite(
            item = item,
            removalId = itemId,
            addFavouriteFood = addFavouriteFood,
            removeFavouriteFood = removeFavouriteFood,
        ) { newValue -> onFavouriteChanged(itemId, newValue) }
    }

    suspend fun onConfirm() {
        if (_state.value.isLoading) return
        if (grams <= 0) {
            alertItem.value = AlertItem(titleRes = R.string.foodQuantity_error_invalidQuantity)
            return
        }
        _state.value = LoadingState.Loading
        try {
            try {
                _mealTypes.value = fetchMealTypes()
            } catch (error: CancellationException) {
                throw error
            } catch (error: Exception) {
                Log.warning(error, Constants.LogCategory.FOOD_QUANTITY)
            }
            val mealTypeId: String?
            if (hasUserSelectedMealType) {
                val selectedId = selectedMealTypeId.value
                if (selectedId == null || _mealTypes.value.none { it.id == selectedId }) {
                    alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
                    return
                }
                mealTypeId = selectedId
            } else {
                mealTypeId = _mealTypes.value.mealType(selectedDate)?.id
            }
            saveFoodConsumed(item, grams = grams, date = selectedDate, mealTypeId = mealTypeId)
            onSaved()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.FOOD_QUANTITY)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    companion object {
        private const val CHECKMARK_DURATION_MILLIS = 2_000L

        fun defaultUnit(item: FoodItemDomain): FoodQuantityUnit =
            item.portions.firstOrNull()?.let { FoodQuantityUnit.Portion(it) } ?: FoodQuantityUnit.Grams
    }
}
